src/main/java/com/minecolonies/coremod/colony/jobs/JobDeliveryman.java

```diff
@@ -87,61 +131,104 @@ public SoundEvent getMoveAwaySound()
+        return !taskQueue.isEmpty() || returning;
+    }
+
+    /**
+     * Returns the {@link IRequest} of the current Task.
+     *
+     * @return {@link IRequest} of the current Task.
+     */
+    public IRequest<? extends Delivery> getCurrentTask()
+    {
+        if (taskQueue.isEmpty())
         {
-            return getCitizen().isFemale() ? DeliverymanSounds.Female.offToBed : null;
+            return null;
         }
-        return null;
+        return getColony().getRequestManager().getRequestForToken(taskQueue.peekFirst());
     }
 
     /**
-     * Generate your AI class to register.
+     * Method used to add a request to the queue
      *
-     * @return your personal AI instance.
+     * @param token The token of the requests to add.
      */
-    @NotNull
-    @Override
-    public AbstractAISkeleton<JobDeliveryman> generateAI()
+    public void addRequest(@NotNull final IToken token)
     {
-        return new EntityAIWorkDeliveryman(this);
+        taskQueue.add(token);
     }
 
     /**
-     * Returns whether or not the job has a destination.
+     * Method called to mark the current request as finished.
      *
-     * @return true if has destination, otherwise false.
+     * @param successful True when the processing was successful, false when not.
      */
-    public boolean hasDestination()
+    public void finishRequest(@NotNull final boolean successful)
     {
-        return destination != null;
+        if (taskQueue.isEmpty())
+        {
+            return;
+        }
+
+        this.setReturning(true);
+        IToken current = taskQueue.removeFirst();
+
+        getColony().getRequestManager().updateRequestState(current, successful ? RequestState.COMPLETED : RequestState.CANCELLED);
     }
 
-    public boolean isNeeded()
+    /**
+     * Called when a task that is being scheduled is being canceled.
+     *
+     * @param token token of the task to be deleted.
+     */
+    public void onTaskDeletion(@NotNull final IToken token)
+    {
+        if (taskQueue.contains(token))
+        {
+            if (taskQueue.peek().equals(token))
+            {
+                this.setReturning(true);
+            }
+
+            taskQueue.remove(token);
+        }
+    }
+
+    /**
+     * Method to get the task queue of this job.
+     *
+     * @return The task queue.
+     */
+    public List<IToken> getTaskQueue()
     {
-        final Colony colony = getCitizen().getColony();
-        return colony != null && !colony.getDeliverymanRequired().isEmpty();
+        return ImmutableList.copyOf(taskQueue);
     }
 
     /**
-     * Returns the {@link BlockPos} of the destination.
+     * Method used to check if this DMan is trying to return to the warehouse to clean up.
      *
-     * @return {@link BlockPos} of the destination.
+     * @return True when this DMan is returning the warehouse to clean his inventory.
      */
-    public BlockPos getDestination()
+    public boolean getReturning()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename this method to start with "is" or "has". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2047)